// ultralytics/dist/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = __importDefault(require("express"));
const morgan_1 = __importDefault(require("morgan"));
const helmet_1 = __importDefault(require("helmet"));
const express_rate_limit_1 = __importDefault(require("express-rate-limit"));
const swagger_ui_express_1 = __importDefault(require("swagger-ui-express"));
const yamljs_1 = __importDefault(require("yamljs"));
const path_1 = __importDefault(require("path"));
const db = __importStar(require("./db"));
const config_1 = __importDefault(require("./config"));
const auth_1 = require("./middleware/auth");
const errorHandler_1 = require("./middleware/errorHandler");
const errors_1 = require("./errors");
const validation_1 = require("./validation");
const metrics_1 = require("./metrics");
const dashboard_1 = __importDefault(require("./routes/dashboard"));
const export_1 = __importDefault(require("./routes/export"));
const analytics_1 = __importDefault(require("./routes/analytics"));
const privacy_1 = __importDefault(require("./routes/privacy"));
const replay_1 = __importDefault(require("./routes/replay"));
// Load OpenAPI specification
const swaggerDocument = yamljs_1.default.load(path_1.default.join(__dirname, '../docs/openapi.yaml'));
const app = (0, express_1.default)();
const PORT = config_1.default.port;
// Parse JSON bodies
app.use(express_1.default.json());
// Security headers
app.use((0, helmet_1.default)({
    contentSecurityPolicy: {
        directives: {
            defaultSrc: ["'self'"],
            styleSrc: ["'self'", "'unsafe-inline'"],
            scriptSrc: ["'self'", "'unsafe-inline'"],
            imgSrc: ["'self'", "data:", "validator.swagger.io"],
        },
    },
}));
// Request logging
app.use((0, morgan_1.default)('combined'));
// Rate limiting
const limiter = (0, express_rate_limit_1.default)({
    windowMs: config_1.default.rateLimit.windowMs,
    max: config_1.default.rateLimit.max,
    message: {
        error: 'Too many requests',
        message: 'Please try again later'
    }
});
app.use('/api', limiter);
// API Documentation
app.use('/docs', swagger_ui_express_1.default.serve, swagger_ui_express_1.default.setup(swaggerDocument, {
    customCss: '.swagger-ui .topbar { display: none }',
    customSiteTitle: 'Ultralytics API Documentation'
}));
// Health check endpoint
app.get('/health', async (_req, res) => {
    const timestamp = new Date().toISOString();
    // Check database connectivity
    let dbStatus = 'unknown';
    let dbLatency = null;
    try {
        const start = Date.now();
        await db.query('SELECT 1');
        dbLatency = Date.now() - start;
        dbStatus = 'connected';
    }
    catch (error) {
        dbStatus = 'disconnected';
        console.error('Database health check failed:', error.message);
    }
    // Get pool stats
    const poolStats = db.getPoolStats();
    const health = {
        status: dbStatus === 'connected' ? 'ok' : 'degraded',
        timestamp: timestamp,
        uptime: process.uptime(),
        database: {
            status: dbStatus,
            latencyMs: dbLatency,
            pool: poolStats
        }
    };
    const statusCode = dbStatus === 'connected' ? 200 : 503;
    res.status(statusCode).json(health);
});
// Kubernetes liveness probe - checks if the process is running
// Returns 200 if the server can respond, regardless of dependencies
app.get('/healthz', (_req, res) => {
    res.status(200).json({
        status: 'ok',
        timestamp: new Date().toISOString()
    });
});
// Kubernetes readiness probe - checks if the server is ready to accept traffic
// Returns 200 only if all dependencies (database) are available
app.get('/readyz', async (_req, res) => {
    try {
        // Check database connection with a timeout
        const timeoutPromise = new Promise((_, reject) => {
            setTimeout(() => reject(new Error('Database check timeout')), 5000);
        });
        const dbCheckPromise = db.query('SELECT 1');
        await Promise.race([dbCheckPromise, timeoutPromise]);
        res.status(200).json({
            status: 'ready',
            timestamp: new Date().toISOString(),
            checks: {
                database: 'ok'
            }
        });
    }
    catch (error) {
        console.error('Readiness check failed:', error.message);
        res.status(503).json({
            status: 'not ready',
            timestamp: new Date().toISOString(),
            checks: {
                database: 'failed'
            },
            error: error.message
        });
    }
});
// Prometheus metrics endpoint
app.get('/metrics', async (_req, res) => {
    try {
        res.set('Content-Type', (0, metrics_1.getContentType)());
        res.send(await (0, metrics_1.getMetrics)());
    }
    catch (error) {
        console.error('Error collecting metrics:', error.message);
        res.status(500).send('Error collecting metrics');
    }
});
// Apply API key authentication to all /api routes
app.use('/api', auth_1.validateApiKey);
// Event tracking endpoint
app.post('/api/events', async (req, res, next) => {
    try {
        const { name, properties, sessionId, userId } = req.body;
        // Schema validation
        const validation = (0, validation_1.validateEventData)(req.body);
        if (!validation.valid) {
            throw new errors_1.ValidationError(validation.errors || 'Validation failed');
        }
        // Store the event with sanitized properties
        const event = {
            name: name,
            properties: (0, validation_1.sanitizeEventProperties)(properties || {}),
            sessionId: sessionId || null,
            userId: userId || null,
            timestamp: new Date()
        };
        const result = await db.storeEvent(event);
        // Track metrics
        metrics_1.eventsTrackedTotal.inc({ event_type: name });
        // Update session tracking
        if (sessionId) {
            await db.updateSession(sessionId);
        }
        res.status(201).json({
            success: true,
            eventId: result.id
        });
    }
    catch (error) {
        next(error);
    }
});
// Batch event tracking endpoint
app.post('/api/events/batch', async (req, res, next) => {
    try {
        const { events } = req.body;
        // Schema validation
        const validation = (0, validation_1.validateBatchEventData)(req.body);
        if (!validation.valid) {
            throw new errors_1.ValidationError(validation.errors || 'Validation failed');
        }
        // Process events in chunks to avoid memory issues with large batches
        const CHUNK_SIZE = 100;
        const results = [];
        const sessionsToUpdate = new Set();
        // Process in chunks for better memory management and reliability
        for (let i = 0; i < events.length; i += CHUNK_SIZE) {
            const chunk = events.slice(i, i + CHUNK_SIZE);
            // Process each event in the chunk
            const chunkPromises = chunk.map(async (eventData) => {
                const event = {
                    name: eventData.name,
                    properties: (0, validation_1.sanitizeEventProperties)(eventData.properties || {}),
                    sessionId: eventData.sessionId || null,
                    userId: eventData.userId || null,
                    timestamp: eventData.timestamp ? new Date(eventData.timestamp) : new Date()
                };
                const result = await db.storeEvent(event);
                if (eventData.sessionId) {
                    sessionsToUpdate.add(eventData.sessionId);
                }
                return result.id;
            });
            // Wait for the chunk to complete before processing the next
            const chunkResults = await Promise.all(chunkPromises);
            results.push(...chunkResults);
        }
        // Update sessions
        for (const sessionId of sessionsToUpdate) {
            await db.updateSession(sessionId);
        }
        // Track metrics
        metrics_1.batchEventsTotal.inc(results.length);
        res.status(201).json({
            success: true,
            eventIds: results,
            count: results.length
        });
    }
    catch (error) {
        next(error);
    }
});
// Query events endpoint
app.get('/api/events', async (req, res, next) => {
    try {
        const { startDate, endDate, name, sessionId, userId } = req.query;
        const limit = req.query.limit ? parseInt(req.query.limit, 10) : 100;
        const offset = req.query.offset ? parseInt(req.query.offset, 10) : 0;
        let queryText = 'SELECT * FROM events WHERE 1=1';
        const params = [];
        let paramCount = 0;
        // Filter by date range
        if (startDate) {
            paramCount++;
            queryText += ` AND timestamp >= $${paramCount}`;
            params.push(new Date(startDate));
        }
        if (endDate) {
            paramCount++;
            queryText += ` AND timestamp <= $${paramCount}`;
            params.push(new Date(endDate));
        }
        // Filter by event name
        if (name) {
            paramCount++;
            queryText += ` AND name = $${paramCount}`;
            params.push(name);
        }
        // Filter by session ID
        if (sessionId) {
            paramCount++;
            queryText += ` AND session_id = $${paramCount}`;
            params.push(sessionId);
        }
        // Filter by user ID
        if (userId) {
            paramCount++;
            queryText += ` AND user_id = $${paramCount}`;
            params.push(userId);
        }
        // Order and pagination
        queryText += ' ORDER BY timestamp DESC';
        paramCount++;
        queryText += ` LIMIT $${paramCount}`;
        params.push(Math.min(limit, 1000)); // Max 1000 results
        paramCount++;
        queryText += ` OFFSET $${paramCount}`;
        params.push(offset);
        const result = await db.query(queryText, params);
        res.json({
            events: result.rows,
            count: result.rows.length,
            offset: offset,
            limit: limit
        });
    }
    catch (error) {
        next(error);
    }
});
// Dashboard routes
app.use('/api/dashboard', dashboard_1.default);
// Export routes
app.use('/api/export', export_1.default);
// Analytics routes
app.use('/api/analytics', analytics_1.default);
// Privacy routes
app.use('/api/privacy', privacy_1.default);
// Replay routes
app.use('/api/replay', replay_1.default);
// 404 handler for unknown routes
app.use(errorHandler_1.notFoundHandler);
// Error handling middleware (must be last)
app.use(errorHandler_1.errorHandler);
// Start server
if (require.main === module) {
    app.listen(PORT, () => {
        console.log(`Ultralytics server running on port ${PORT}`);
    });
}
exports.default = app;
